### PR TITLE
Add support for logging markers as tags

### PR DESCRIPTION
**Pull request by joschi**
Logback (SLF4J) and Log4j 2 support Markers, which can contain useful information.

This change adds support for optionally logging these markers as tags in the created/encoded log messages.

By default, this feature is disabled and can be enabled by setting the `includeMarkers` configuration setting to `true` in the Log4j 2 `EcsLayout` and Logback `EcsEncoder`.

See also:
* https://logging.apache.org/log4j/2.0/manual/markers.html
* https://www.slf4j.org/api/org/slf4j/Marker.html

### CODE REVIEW

**Comment by felixbarny** [comment]
Great, thanks a lot for the contribution!
I must say I don't have lots of experience with markers. Do you think it may be a good fit for [`event.dataset`](https://www.elastic.co/guide/en/ecs/current/ecs-event.html)? Or would `tags` be more appropriate in your opinion?

I think there is also some overlap with #12

**Comment by joschi** [comment]
>  I must say I don't have lots of experience with markers. Do you think it may be a good fit for `event.dataset`? Or would `tags` be more appropriate in your opinion?

I think `tags` is a better match since there can be multiple markers per message and `event.dataset` is a scalar value, if I see that correctly.

Furthermore, if we take the example from the [Log4j 2 documentation](https://logging.apache.org/log4j/2.0/manual/markers.html) about markers, one could search for all SQL queries ("SQL") or only UPDATE statements ("SQL_UPDATE") in the log messages. When using `event.dataset`, this wouldn't be easily possible I think.

```java
private static final Marker SQL_MARKER = MarkerManager.getMarker("SQL");
private static final Marker UPDATE_MARKER = MarkerManager.getMarker("SQL_UPDATE").setParents(SQL_MARKER);
private static final Marker QUERY_MARKER = MarkerManager.getMarker("SQL_QUERY").setParents(SQL_MARKER);
```

> I think there is also some overlap with #12

Regarding the integration tests, yes.

**Comment by ruflin** [comment]
+1 on using tags. I guess it then would look like: `tags: ["SQL", "SQL_UPDATE"]` as an example?

**Comment by joschi** [comment]
@ruflin Yes, correct.

In case of Log4j 2 it would be `tags: ["SQL_UPDATE", "SQL"]` and in Logback it would be `tags: ["SQL", "SQL_UPDATE"]` because of how the direction of the parent-child relation is modelled in the respective logging framework. 😉

File: log4j2-ecs-layout/src/main/java/co/elastic/logging/log4j2/EcsLayout.java [hunk 1]
```diff
@@ -152,16 +156,35 @@ private void serializeLabels(LogEvent event, StringBuilder builder) {
 
     private void serializeTags(LogEvent event, StringBuilder builder) {
         List<String> contextStack = event.getContextStack().asList();
+        Marker marker = event.getMarker();
+        if (!contextStack.isEmpty() || (includeMarkers && marker != null)) {
```
Comment (felixbarny) [inline log4j2-ecs-layout/src/main/java/co/elastic/logging/log4j2/EcsLayout.java]:
```suggestion
        boolean hasTags = !contextStack.isEmpty() || (includeMarkers && marker != null);
        if (hasTags) {
```

File: log4j2-ecs-layout/src/main/java/co/elastic/logging/log4j2/EcsLayout.java [hunk 2]
```diff
@@ -152,16 +156,35 @@ private void serializeLabels(LogEvent event, StringBuilder builder) {
...
-            // removes last comma
-            builder.setLength(builder.length() - 1);
-            builder.append("],");
+        }
+
+        if (includeMarkers && marker != null) {
+            serializeMarker(builder, marker);
+        }
+
+        if (!contextStack.isEmpty() || (includeMarkers && marker != null)) {
```
Comment (felixbarny) [inline log4j2-ecs-layout/src/main/java/co/elastic/logging/log4j2/EcsLayout.java]:
```suggestion
        if (hasTags) {
```

File: log4j2-ecs-layout/src/test/java/co/elastic/logging/log4j2/Log4j2EcsLayoutIntegrationTest.java [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+
+import java.io.IOException;
+import java.net.URISyntaxException;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+class Log4j2EcsLayoutIntegrationTest extends AbstractEcsLoggingTest {
```
Comment (felixbarny) [inline log4j2-ecs-layout/src/test/java/co/elastic/logging/log4j2/Log4j2EcsLayoutIntegrationTest.java]: There's some duplication from the `Log4j2EcsLayoutTest`, maybe extract a common superclass?

**Comment by felixbarny** [comment]
@pgomulka Would mapping markers to `tags` also work for Elasticsearch?

**Comment by felixbarny** [comment]
Jenkins run the tests please

**Comment by felixbarny** [comment]
@joschi There seem to be some failing tests: https://apm-ci.elastic.co/blue/organizations/jenkins/apm-agent-java%2Fjava-ecs-logging-mbp/detail/PR-14/7/pipeline/

**Comment by pgomulka** [comment]
> @pgomulka Would mapping markers to tags also work for Elasticsearch?

I think this should be fine, I was aiming to refactor this way. But ES is using markers the way they should be - we just treat this basically as a label.

**Comment by joschi** [comment]
@felixbarny Unfortunately I'm unable to see the Jenkins build output at https://apm-ci.elastic.co/blue/organizations/jenkins/apm-agent-java%2Fjava-ecs-logging-mbp/detail/PR-14/7/pipeline/

**Comment by felixbarny** [comment]
This is the stack trace:
```
java.lang.NullPointerException
	at co.elastic.logging.log4j2.Log4j2EcsLayoutIntegrationTest.getLastLogLine(Log4j2EcsLayoutIntegrationTest.java:77)
	at co.elastic.logging.log4j2.AbstractLog4j2EcsLayoutTest.testMarker(AbstractLog4j2EcsLayoutTest.java:67)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.junit.platform.commons.util.ReflectionUtils.invokeMethod(ReflectionUtils.java:675)
	at org.junit.jupiter.engine.execution.MethodInvocation.proceed(MethodInvocation.java:60)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain$ValidatingInvocation.proceed(InvocationInterceptorChain.java:125)
	at org.junit.jupiter.engine.extension.TimeoutExtension.intercept(TimeoutExtension.java:132)
	at org.junit.jupiter.engine.extension.TimeoutExtension.interceptTestableMethod(TimeoutExtension.java:124)
	at org.junit.jupiter.engine.extension.TimeoutExtension.interceptTestMethod(TimeoutExtension.java:74)
	at org.junit.jupiter.engine.execution.ExecutableInvoker$ReflectiveInterceptorCall.lambda$ofVoidMethod$0(ExecutableInvoker.java:115)
	at org.junit.jupiter.engine.execution.ExecutableInvoker.lambda$invoke$0(ExecutableInvoker.java:105)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain$InterceptedInvocation.proceed(InvocationInterceptorChain.java:104)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.proceed(InvocationInterceptorChain.java:62)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.chainAndInvoke(InvocationInterceptorChain.java:43)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.invoke(InvocationInterceptorChain.java:35)
	at org.junit.jupiter.engine.execution.ExecutableInvoker.invoke(ExecutableInvoker.java:104)
	at org.junit.jupiter.engine.execution.ExecutableInvoker.invoke(ExecutableInvoker.java:98)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.lambda$invokeTestMethod$6(TestMethodTestDescriptor.java:202)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.invokeTestMethod(TestMethodTestDescriptor.java:198)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.execute(TestMethodTestDescriptor.java:135)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.execute(TestMethodTestDescriptor.java:69)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$5(NodeTestTask.java:135)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$7(NodeTestTask.java:125)
	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:135)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:123)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:122)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:80)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1540)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:38)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$5(NodeTestTask.java:139)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$7(NodeTestTask.java:125)
	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:135)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:123)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:122)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:80)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1540)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:38)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$5(NodeTestTask.java:139)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$7(NodeTestTask.java:125)
	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:135)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:123)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:122)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:80)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.submit(SameThreadHierarchicalTestExecutorService.java:32)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor.execute(HierarchicalTestExecutor.java:57)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestEngine.execute(HierarchicalTestEngine.java:51)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:170)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:154)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:90)
	at org.junit.platform.surefire.provider.JUnitPlatformProvider.invokeAllTests(JUnitPlatformProvider.java:132)
	at org.junit.platform.surefire.provider.JUnitPlatformProvider.invoke(JUnitPlatformProvider.java:111)
	at org.apache.maven.surefire.booter.ForkedBooter.invokeProviderInSameClassLoader(ForkedBooter.java:290)
	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:242)
	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:121)
	Suppressed: java.lang.NullPointerException
		at co.elastic.logging.log4j2.Log4j2EcsLayoutIntegrationTest.tearDown(Log4j2EcsLayoutIntegrationTest.java:72)
		at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
		at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
		at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
		at java.base/java.lang.reflect.Method.invoke(Method.java:566)
		at org.junit.platform.commons.util.ReflectionUtils.invokeMethod(ReflectionUtils.java:675)
		at org.junit.jupiter.engine.execution.MethodInvocation.proceed(MethodInvocation.java:60)
		at org.junit.jupiter.engine.execution.InvocationInterceptorChain$ValidatingInvocation.proceed(InvocationInterceptorChain.java:125)
		at org.junit.jupiter.engine.extension.TimeoutExtension.intercept(TimeoutExtension.java:132)
		at org.junit.jupiter.engine.extension.TimeoutExtension.interceptLifecycleMethod(TimeoutExtension.java:111)
		at org.junit.jupiter.engine.extension.TimeoutExtension.interceptAfterEachMethod(TimeoutExtension.java:95)
		at org.junit.jupiter.engine.execution.ExecutableInvoker$ReflectiveInterceptorCall.lambda$ofVoidMethod$0(ExecutableInvoker.java:115)
		at org.junit.jupiter.engine.execution.ExecutableInvoker.lambda$invoke$0(ExecutableInvoker.java:105)
		at org.junit.jupiter.engine.execution.InvocationInterceptorChain$InterceptedInvocation.proceed(InvocationInterceptorChain.java:104)
		at org.junit.jupiter.engine.execution.InvocationInterceptorChain.proceed(InvocationInterceptorChain.java:62)
		at org.junit.jupiter.engine.execution.InvocationInterceptorChain.chainAndInvoke(InvocationInterceptorChain.java:43)
		at org.junit.jupiter.engine.execution.InvocationInterceptorChain.invoke(InvocationInterceptorChain.java:35)
		at org.junit.jupiter.engine.execution.ExecutableInvoker.invoke(ExecutableInvoker.java:104)
		at org.junit.jupiter.engine.execution.ExecutableInvoker.invoke(ExecutableInvoker.java:98)
		at org.junit.jupiter.engine.descriptor.ClassBasedTestDescriptor.invokeMethodInExtensionContext(ClassBasedTestDescriptor.java:464)
		at org.junit.jupiter.engine.descriptor.ClassBasedTestDescriptor.lambda$synthesizeAfterEachMethodAdapter$17(ClassBasedTestDescriptor.java:454)
		at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.lambda$invokeAfterEachMethods$9(TestMethodTestDescriptor.java:228)
		at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.lambda$invokeAllAfterMethodsOrCallbacks$12(TestMethodTestDescriptor.java:256)
		at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
		at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.lambda$invokeAllAfterMethodsOrCallbacks$13(TestMethodTestDescriptor.java:256)
		at java.base/java.util.ArrayList.forEach(ArrayList.java:1540)
		at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.invokeAllAfterMethodsOrCallbacks(TestMethodTestDescriptor.java:255)
		at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.invokeAfterEachMethods(TestMethodTestDescriptor.java:226)
		at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.execute(TestMethodTestDescriptor.java:139)
		... 40 more
```

I don't know where this is coming from.

Could you [allow edits from maintainers](https://help.github.com/en/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork)?

If you add `assertThat(root.getAppenders()).containsKey("TestAppender");` to the `@BeforeEach` method in `Log4j2EcsLayoutIntegrationTest` we might find out more.

**Comment by joschi** [comment]
After some playing around with the URL, I've found https://apm-ci.elastic.co/job/apm-agent-java/job/java-ecs-logging-mbp/view/change-requests/job/PR-14/7.

@felixbarny I think the issue was that both Log4j 2 tests, when run in parallel, modified the same singleton LoggerContext which made at least one of them fail.

> Could you allow edits from maintainers?

That's already enabled. At least the box is checked.

**Comment by felixbarny** [comment]
Heh, I didn’t know they were running in parallel :smile:

**Pull request merged**